links/links.go

```diff
@@ -0,0 +1,214 @@
+	"APP_RE_OPEN":    AppReOpen,
+}
+
+func (e EventType) String() string {
+	return eventTypesByID[e]
+}
+
+// MarshalJSON makes the "enum" usable in marshalling json
+func (e *EventType) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + eventTypesByID[*e] + `"`), nil
+}
+
+// UnmarshalJSON is used to read the json files into the "enum"
+func (e *EventType) UnmarshalJSON(b []byte) error {
+	var s string
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+	var ok bool
+	if *e, ok = eventTypesByName[s]; !ok {
+		return fmt.Errorf("unknown event type %q", s)
+	}
+	return nil
+}
+
+// StatOptions are used in the request for GetLinkStats, it is used to request data
+// going back DurationDays days.
+type StatOptions struct {
+	DurationDays int
+}
+
+// Client is the interface for the Firebase dynamics links service.
+//
+// Client is the entry point to the dynamic links functions
+type Client struct {
+	hc                *internal.HTTPClient
+	linksEndPoint     string
+	linksStatsRequest string
+}
+
+// NewClient creates a new instance of the Firebase Auth Client.
+//
+// This function can only be invoked from within the SDK. Client applications should access the
+// Auth service through firebase.App.
+func NewClient(ctx context.Context, opts ...option.ClientOption) (*Client, error) {
+	hc, _, err := transport.NewHTTPClient(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		hc:                &internal.HTTPClient{Client: hc},
+		linksEndPoint:     linksEndPoint,
+		linksStatsRequest: linksStatsRequest,
+	}, nil
+}
+
+// LinkStats returns the link stats given a url, and the duration (inside the StatOptions)
+// Returns a LinkStats object which contains a list of EventStats.
+// The service account with which the firebase_app is validated must be associated with the project
+// for which the stats are requested.
+// If the URI prefix for the shortlink belongs to the project but the link suffix has either not
+// been created or has no data in the requested period, the LinkStats object will contain an
+// empty list of EventStats.
+func (c *Client) LinkStats(ctx context.Context, shortLink string, statOptions StatOptions) (*LinkStats, error) {
+	if ok := strings.HasPrefix(shortLink, "https://"); !ok {
+		return nil, fmt.Errorf("short link must start with `https://`")
+	}
+	if ok := statOptions.DurationDays > 0; !ok {
+		return nil, fmt.Errorf("durationDays must be > 0")
+	}
+	request := &internal.Request{
+		Method: http.MethodGet,
+		URL:    c.makeURLForLinkStats(shortLink, statOptions),
+	}
+
+	resp, err := c.hc.Do(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	if err := resp.CheckStatus(http.StatusOK); err != nil {
+		return nil, err
+	}
+	var result LinkStats
+	err = json.Unmarshal(resp.Body, &result)
+	return &result, err
+}
+
+func (c *Client) makeURLForLinkStats(shortLink string, statOptions StatOptions) string {
+	return fmt.Sprintf(c.linksEndPoint+"/"+c.linksStatsRequest,
```

Comment:
In general, it's a bad idea to use a variable for a format string. If these were the constants above, then that would be fine. But I'm concerned about how these might have different values.

Maybe it would make sense to use go's URL type here instead of doing string manipulation?

Reply:
the different values would be needed for tests.

Reply:
That seems to be not the best solution, replacing the endpoint for Host, 
and moving the '/v1/' to the request? 

would you rather I encapsulate this as a formatting function ?